### PR TITLE
Spread transform/traverse/mapOver into virtual methods of Tree/Type

### PR DESCRIPTION
**Pull request by retronym**
Analysis of JIT logs shows that a handful of very hot pattern matches in tree and type traversal run through ~10 deoptimizations before finding a steady state. Even once JITted, they show up in profiles as consuming a few percent of time, even though they are mostly just doing dispatching.

In this PR, I have changed them to virtual calls. Analysis of the JIT logs shows better stability for these methods, and the benchmarks suggest that the compiler has better peak performance.

### CODE REVIEW

**Comment by retronym** [comment]
Benchmarks scheduled to [appear soon](https://scala-ci.typesafe.com/grafana/dashboard/db/pr-validation?var-scalaSha=6d29764%7C556a56b897%7Cac0d7f0680%7C72c8868f18%7C6d2a6c2c1a%7Cb1f0a40f79&var-source=All&var-bench=HotScalacBenchmark.compile&var-host=scalabench@scalabench@&from=1349846730990&to=1504675530990). I measured approx 4% speedup for the `scalap` benchmark locally.

**Comment by retronym** [comment]
Reviving #5906, although I'd forgotten about the test failures that need investigation ☹️

File: src/reflect/scala/reflect/internal/Types.scala [hunk 1]
```diff
@@ -1129,6 +1131,7 @@ trait Types
     override def safeToString: String = "<error>"
     override def narrow: Type = this
     override def kind = "ErrorType"
+    override def mapOver(map: TypeMap): Type = this
```
Comment (lrytz) [inline src/reflect/scala/reflect/internal/Types.scala]: There are many `override def mapOver(map: TypeMap): Type = this` added in this commit, and they are unnecessary in principle. I suggest to either make `mapOver` in `Type` abstract or throw, or remove all redundant overrides.

**Review by lrytz — COMMENTED** [review]
Looks good! Can you provide some motivation in the message of the commit that reverts c2dc346?

**Comment by sjrd** [comment]
I see a lot of
```
-          super.transform(tree)
+          tree.transform(this)
```
in transforms. Are these mandatory with the new scheme? Will this likely break compiler plugins?

**Comment by lrytz** [comment]
They continue to work, transformers still have a `transform` method that forwards to `tree.transform`.

**Comment by sjrd** [comment]
Cool thanks :)

**Comment by retronym** [comment]
> Looks good! Can you provide some motivation in the message of the commit that reverts c2dc346?

I was aiming for cleaner stack traces in the transformers in order to make it a little easier to navigate through profiles. It didn't seem to make a difference to performance.

File: src/reflect/scala/reflect/internal/Types.scala (hunk 1 above)
Comment (retronym) [reply src/reflect/scala/reflect/internal/Types.scala]: Hmm, I can't remember the sequence of how I arrived at this. I do remember running into problems when `Type` became a SAM and the inliner (prior to #5895) was overzealous. Maybe I made this non-abstract as a workaround.

**Comment by retronym** [comment]
Anything more needed by the reviewers?

**Comment by adriaanm** [comment]
There's a (trivial) conflict. I resolved and moved the revert to the last commit to make it easier to adapt the commit message. I didn't want to push -f to your branch. My branch is at https://github.com/adriaanm/scala/tree/faster/virtual-transform-minimal

**Comment by adriaanm** [comment]
After that, LGTM

**Comment by retronym** [comment]
Added a merge commit to resolve the conflict with #6089

**Comment by retronym** [comment]
Something was wonky with Jenkins in https://scala-ci.typesafe.com/job/scala-2.13.x-validate-main/567/console, the build flow claimed it was waiting for the integrate-ide job, but that has already finished.

**Comment by retronym** [comment]
Merging on Adriaan's conditional LGTM.

**Pull request merged**